R/visualize_embeddings.R

```diff
@@ -143,77 +144,150 @@ keep_tokens <- function(embedding_df, tokens = "[CLS]") {
-        ggplot2::scale_color_manual(values = class_colors,
-                           name = color_field) +
-        ggplot2::geom_text(vjust = 0, nudge_y = 0.5) +
-        ggplot2::geom_point()
-    if (num_class <= 1) {
-        ggp <- ggp + ggplot2::theme(legend.position = "none")
-    }
-    return(ggp)
+  pca_df <- do_pca(embedding_df = embedding_df,
+                   project_vectors = project_vectors,
+                   color_field = color_field,
+                   disambiguate_tokens = disambiguate_tokens)
+  return(
+    plot_pca(pca_df = pca_df,
+             color_field = color_field)
+  )
 }
 
+# do_pca -------------------------------------------------------------
 
+#' Do PCA on Embeddings
+#'
+#' Calculate the 2D PCA for a collection of embedding vectors, in preparation
+#' for plotting.
+#'
+#' @inheritParams display_pca
+#'
+#' @return A tbl_df of the embedding vectors projected onto two principal
+#'   axes.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # assuming something like the following has been run:
+#' # feats <- RBERT::extract_features(...) # See RBERT documentation
+#' # Then:
+#' pca_df <- feats$output %>%
+#'     filter_layer_embeddings(layer_indices = 12L) %>%
+#'     keep_tokens("[CLS]") %>%
+#'     do_pca()
+#' }
+do_pca <- function(embedding_df,
+                   project_vectors = embedding_df,
+                   color_field = NULL,
+                   disambiguate_tokens = TRUE) {
+  num_rows <- nrow(embedding_df)
+  num_rows_proj <- nrow(project_vectors)
+  if (num_rows_proj < 3) {
+    stop("At least three vectors are required for a ",
+         "meaningful PCA plot.")
+  }
+  # Use just the indicated vectors to select the PCA projection.
+  proj_mat <- as.matrix(
+    dplyr::select(project_vectors,
+                  dplyr::matches("V[0-9]+"))
+  )
+  pcs <- stats::prcomp(proj_mat,
+                       retx = TRUE, center = TRUE, scale. = TRUE, rank. = 2L)
+
+  if (disambiguate_tokens) {
+    embedding_df <- dplyr::mutate(embedding_df,
+                                  token = paste(token,
+                                                sequence_index,
+                                                token_index,
+                                                sep = "."))
+  }
+  # Keep all the non-vector columns, for possible use in plotting.
+  tok_labels <- dplyr::select(embedding_df,
+                              -dplyr::matches("V[0-9]+"))
+
+  vec_mat <- as.matrix(
+    dplyr::select(embedding_df,
+                  dplyr::matches("V[0-9]+"))
+  )
+  # Instead of doing PCA here, apply the scaling and projection manually.
+  vec_mat <- scale(vec_mat, center = pcs$center, scale = pcs$scale)
+  projected <- vec_mat %*% pcs$rotation
+
+  pca_df <- dplyr::bind_cols(tok_labels, tibble::as_tibble(projected))
+  return(pca_df)
+}
+
+# plot_pca -------------------------------------------------------------
+
+#' Plot PCA
+#'
+#' Given the output of \code{prep_pca}, make the plot object.
```

Comment:
This definitely isn't `prep_pca` anymore; fix once the name of the other function is solidified.

Reply:
good catch!